File: p131_palindrome partition.py
# Given a string s, partition s such that every substring of the partition is a palindrome.

# Return all possible palindrome partitioning of s.

# Example:

# Input: "aab"
# Output:
# [
#   ["aa","b"],
#   ["a","a","b"]
# ]



class Solution:
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        def helper(s, path):
            if s == '':
                self.res += path[:],
                # if no [:], path is by reference, will pop later
                return
            for i in range(1, len(s)+1):
                if s[:i] == s[:i][::-1]:
                    path += s[:i],
                    helper(s[i:], path)
                    path.pop() # explicit make it backtracking
                    
        self.res = []
        helper(s, [])
        return self.res
        
        

File: p131_palindrome partition.cpp
/*
Given a string s, partition s such that every substring of the partition is a palindrome.

Return all possible palindrome partitioning of s.

Example:

Input: "aab"
Output:
[
  ["aa","b"],
  ["a","a","b"]
]
*/



class Solution {
    vector<vector<string>> res;
public:
    vector<vector<string>> partition(string s) {
        helper(s, {}, 0);
        return res;
    }
    
    void helper(string& s, vector<string> path, int idx) {
        if (idx == s.size()) {
            res.push_back(path);
            return; 
        }
        for (int i = idx; i < s.size(); i++) {
            if (isPa(s, idx, i)) {
                path.push_back(s.substr(idx, i-idx+1));
                helper(s, path, i+1);
                path.pop_back();
            }
        }
    }
    
    bool isPa(string& s, int i, int j) {
        while (i < j && s[i] == s[j]) i++, j--;
        return i >= j;
    }
};
        
        